src/Vat.sol

```diff
@@ -243,55 +243,61 @@ contract Vat {
         require(either(dart >= 0, wish(w, msg.sender)), "Vat/not-allowed-w");
 
         // urn has no debt, or a non-dusty amount
-        require(either(urn.art == 0, tab >= ilk.dust), "Vat/dust");
+        require(either(urn.art == 0, tab >= ilks[i].dust), "Vat/dust");
 
         gem[i][v] = _sub(gem[i][v], dink);
         dai[w]    = _add(dai[w],    dtab);
 
-        urns[i][u] = urn;
-        ilks[i]    = ilk;
+        debt = debt_;
+
+        urns[i][u]  = urn;
+        ilks[i].Art = Art_;
 
         emit Frob(i, u, v, w, dink, dart);
     }
 
     // --- CDP Fungibility ---
     function fork(bytes32 ilk, address src, address dst, int256 dink, int256 dart) external {
-        Urn storage u = urns[ilk][src];
-        Urn storage v = urns[ilk][dst];
-        Ilk storage i = ilks[ilk];
+        Urn memory u = urns[ilk][src];
+        Urn memory v = urns[ilk][dst];
+        uint256 rate_ = ilks[ilk].rate;
+        uint256 spot_ = ilks[ilk].spot;
+        uint256 dust_ = ilks[ilk].dust;
 
         u.ink = _sub(u.ink, dink);
         u.art = _sub(u.art, dart);
         v.ink = _add(v.ink, dink);
         v.art = _add(v.art, dart);
 
-        uint256 utab = u.art * i.rate;
-        uint256 vtab = v.art * i.rate;
+        uint256 utab = u.art * rate_;
+        uint256 vtab = v.art * rate_;
 
         // both sides consent
         require(both(wish(src, msg.sender), wish(dst, msg.sender)), "Vat/not-allowed");
 
         // both sides safe
-        require(utab <= u.ink * i.spot, "Vat/not-safe-src");
-        require(vtab <= v.ink * i.spot, "Vat/not-safe-dst");
+        require(utab <= u.ink * spot_, "Vat/not-safe-src");
+        require(vtab <= v.ink * spot_, "Vat/not-safe-dst");
 
         // both sides non-dusty
-        require(either(utab >= i.dust, u.art == 0), "Vat/dust-src");
-        require(either(vtab >= i.dust, v.art == 0), "Vat/dust-dst");
+        require(either(utab >= dust_, u.art == 0), "Vat/dust-src");
+        require(either(vtab >= dust_, v.art == 0), "Vat/dust-dst");
+
+        urns[ilk][src] = u;
+        urns[ilk][src] = v;
```

Comment:
Should be writing to `dst` urn. Why don't tests pick this up?

Reply:
Yeah good question, I'll look into it.

Reply:
The tests were horrible (they are a bit better now). There was also another critical bug when forking to itself (both ink and art got duplicated).
We have to go through the complete test suite and see what is missing, but these are issues that Certora specs would have picked up as well or at least they would have to (I'll double check that).